pkg/serverless/invocationlifecycle/trace.go

```diff
@@ -39,24 +41,36 @@ var currentExecutionInfo executionStartInfo
 
 // startExecutionSpan records information from the start of the invocation.
 // It should be called at the start of the invocation.
-func startExecutionSpan(startTime time.Time, rawPayload string) {
+func startExecutionSpan(startTime time.Time, rawPayload string, inferredSpan inferredspan.InferredSpan) {
 	currentExecutionInfo.startTime = startTime
-	currentExecutionInfo.traceID = random.Uint64()
-	currentExecutionInfo.spanID = random.Uint64()
+	currentExecutionInfo.traceID = rand.Random.Uint64()
+	currentExecutionInfo.spanID = rand.Random.Uint64()
 	currentExecutionInfo.parentID = 0
 
 	payload := convertRawPayload(rawPayload)
 
+	if inferredspan.IsExpected() {
+		currentExecutionInfo.traceID = inferredSpan.Span.TraceID
+		currentExecutionInfo.parentID = inferredSpan.Span.SpanID
+	}
+
 	if payload.Headers != nil {
 		traceID, e1 := convertStrToUnit64(payload.Headers[TraceIDHeader])
-		parentID, e2 := convertStrToUnit64(payload.Headers[parentIDHeader])
+		parentID, e2 := convertStrToUnit64(payload.Headers[ParentIDHeader])
 
 		if e1 == nil {
 			currentExecutionInfo.traceID = traceID
+			if inferredspan.IsExpected() {
+				inferredSpan.Span.TraceID = traceID
+			}
```

Comment:
I think you could refactor this function to be simpler. Right now, it looks like if inferred spans are enabled, you:

- Set the current execution context to the inferred span
- Overwrite it from the  headers
- Overwrite it back to what it was

Maybe there is way to reorganize this to be a bit cleaner/simpler?